### PR TITLE
[frax-finance-with-delegations] Frax Finance Delegation

### PR DESCRIPTION
**Pull request by FortisFortuna**
Adds delegation capabilities

### CODE REVIEW

File: src/strategies/frax-finance-with-delegations/index.ts [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    provider,
+    addresses,
+    options,
+    snapshot
+  );
+  console.debug('Direct scores', directScores);
+
+  // Calculate the score
+  return Object.fromEntries(
+    addresses.map((address, addr_idx) => {
```
Comment (ChaituVR) [inline src/strategies/frax-finance-with-delegations/index.ts:50]:
```suggestion
    addresses.map(address => {
```

**Comment by bonustrack** [comment]
@FortisFortuna Any reason you don't use the strategy "delegation" directly?

**Comment by FortisFortuna** [comment]
> @FortisFortuna Any reason you don't use the strategy "delegation" directly?

Our vote weight is based on a few things:
1) Simple ERC20 style balance of FXS token.
2) veFXS balance (staked FXS, analogous to veCRV)
3) The FXS value of the UniV2 FRAX/FXS LP token in a specific farm
4) (New) Delegated 1,2,3 scores from another address

**Comment by bonustrack** [comment]
@FortisFortuna You can do the same with the strategy delegation, on this strategy you can define 1,2,3 strategies to be counted

**Comment by FortisFortuna** [comment]
@bonustrack @ChaituVR When I tried that, the tests failed because it was a catch-22 situation because no delegations currently exist, even though I tried doing it with https://snapshot.org/#/delegate

**Comment by ChaituVR** [comment]
> @bonustrack @ChaituVR When I tried that, the tests failed because it was a catch-22 situation because no delegations currently exist, even though I tried doing it with https://snapshot.org/#/delegate

@FortisFortuna What went wrong when you tried delegation? 🤔 

Try using delegation strategy in your space to pass different strategies into it, https://snapshot.org/#/strategy/delegation

**Comment by FortisFortuna** [comment]
So this should work?

{
  "symbol": "Frax Finance (delegated)",
  "strategies": [
    {
      "name": "frax-finance",
      "params": {
        "symbol": "FXS",
        "decimals": 18,
        "FXS": "0x3432B6A60D23Ca0dFCa7761B7ab56459D9C964D0",
        "VEFXS": "0xc8418aF6358FFddA74e09Ca9CC3Fe03Ca6aDC5b0",
        "UNI_LP_FRAX_FXS": "0xE1573B9D29e2183B1AF0e743Dc2754979A40D237",
        "FARMING_UNI_LP_FRAX_FXS": "0xda2c338350a0E59Ce71CDCED9679A3A590Dd9BEC"
      }
    }
  ]
}

**Comment by FortisFortuna** [comment]
I got it to work, but it seems I would also need to delegate to myself as well right?
Assume two voters, A and B
A has 1000, B has 500.
A then delegates to B.

Right now, B is showing 1000 instead of the sum, which should be 500 + 1000 = 1500, unless I am doing something wrong.

**Comment by ChaituVR** [comment]
> Assume two voters, A and B
A has 1000, B has 500.
A then delegates to B.
> Right now, B is showing 1000 instead of the sum, which should be 500 + 1000 = 1500, unless I am doing something wrong.

Ya you are correct It will return only delegated voting power, to add up original voting power too, you may have to add another strategy to get non delegated voting power, for example: https://hub.snapshot.org/api/spaces/balancer.eth they check both delegated and not delegated balance to vote

So in case both A and B vote, then the delegated voting power is not calculated for B. Only the vote of A will be calculated. The vote of B will be counted if A does not vote.

**Comment by ChaituVR** [comment]
Here is an example curl request you can play with:
```
curl --location --request POST 'https://score.snapshot.org/api/scores' \
--header 'authority: score.snapshot.org' \
--header 'sec-ch-ua: " Not A;Brand";v="99", "Chromium";v="96", "Google Chrome";v="96"' \
--header 'sec-ch-ua-mobile: ?0' \
--header 'user-agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/96.0.4664.110 Safari/537.36' \
--header 'sec-ch-ua-platform: "Windows"' \
--header 'content-type: application/json' \
--header 'accept: */*' \
--header 'origin: https://snapshot.org' \
--header 'sec-fetch-site: same-site' \
--header 'sec-fetch-mode: cors' \
--header 'sec-fetch-dest: empty' \
--header 'referer: https://snapshot.org/' \
--header 'accept-language: en-US,en;q=0.9' \
--data-raw '{
    "params": {
        "space": "balancer.eth",
        "network": "1",
        "snapshot": 13843331,
        "strategies": [
            {
                "name": "delegation",
                "params": {
                    "symbol": "BAL & BPT multichain (delegated)",
                    "strategies": [
                        {
                            "name": "multichain",
                            "params": {
                                "strategies": [
                                    {
                                        "name": "erc20-balance-of",
                                        "params": {
                                            "address": "0xba100000625a3754423978a60c9317c58a424e3D",
                                            "decimals": 18
                                        },
                                        "network": "1"
                                    }
                                ],
                                "startBlocks": {
                                    "1": 0,
                                    "137": 12685221,
                                    "42161": 2178873
                                }
                            }
                        },
                        {
                            "name": "multichain",
                            "params": {
                                "strategies": [
                                    {
                                        "name": "balancer",
                                        "params": {
                                            "symbol": "BAL BPT V1+V2",
                                            "address": "0xba100000625a3754423978a60c9317c58a424e3D",
                                            "decimals": 18
                                        },
                                        "network": "1"
                                    },
                                    {
                                        "name": "balancer",
                                        "params": {
                                            "symbol": "BAL BPT V1+V2",
                                            "address": "0x9a71012B13CA4d3D0Cdc72A177DF3ef03b0E76A3",
                                            "decimals": 18
                                        },
                                        "network": "137"
                                    },
                                    {
                                        "name": "balancer",
                                        "params": {
                                            "symbol": "BAL BPT V1+V2",
                                            "address": "0x040d1EdC9569d4Bab2D15287Dc5A4F10F56a56B8",
                                            "decimals": 18
                                        },
                                        "network": "42161"
                                    }
                                ],
                                "startBlocks": {
                                    "1": 0,
                                    "137": 12685221,
                                    "42161": 2178873
                                }
                            }
                        }
                    ]
                }
            },
            {
                "name": "multichain",
                "params": {
                    "symbol": "BPT (multichain)",
                    "strategies": [
                        {
                            "name": "balancer",
                            "params": {
                                "symbol": "BAL BPT V1+V2",
                                "address": "0xba100000625a3754423978a60c9317c58a424e3D",
                                "decimals": 18
                            },
                            "network": "1"
                        },
                        {
                            "name": "balancer",
                            "params": {
                                "symbol": "BAL BPT V1+V2",
                                "address": "0x9a71012B13CA4d3D0Cdc72A177DF3ef03b0E76A3",
                                "decimals": 18
                            },
                            "network": "137"
                        },
                        {
                            "name": "balancer",
                            "params": {
                                "symbol": "BAL BPT V1+V2",
                                "address": "0x040d1EdC9569d4Bab2D15287Dc5A4F10F56a56B8",
                                "decimals": 18
                            },
                            "network": "42161"
                        }
                    ],
                    "startBlocks": {
                        "1": 0,
                        "137": 12685221,
                        "42161": 2178873
                    }
                }
            }
        ],
        "addresses": [
            "0xb2a33ae0e07fd2ca8dbde9545f6ce0b3234dc4e8",
            "0x67a57535b11445506a9e340662cd0c9755e5b1b4"
        ]
    }
}'
```

Try removing `0x67a57535b11445506a9e340662cd0c9755e5b1b4` from addresses (if A doesn't vote)

**Comment by FortisFortuna** [comment]
`{
    "params": {
        "space": "frax.eth",
        "network": "1",
        "snapshot": 13936442,
        "strategies": [
            {
                "name": "delegation",
                "params": {
                    "symbol": "Frax Finance (delegated)",
                    "strategies": [
                        {
                            "name": "frax-finance",
                            "params": {
                                "symbol": "FXS",
                                "decimals": 18,
                                "FXS": "0x3432B6A60D23Ca0dFCa7761B7ab56459D9C964D0",
                                "VEFXS": "0xc8418aF6358FFddA74e09Ca9CC3Fe03Ca6aDC5b0",
                                "UNI_LP_FRAX_FXS": "0xE1573B9D29e2183B1AF0e743Dc2754979A40D237",
                                "FARMING_UNI_LP_FRAX_FXS": "0xda2c338350a0E59Ce71CDCED9679A3A590Dd9BEC"
                            }
                        }
                    ]
                }
            },
            {
                "name": "frax-finance",
                "params": {
                    "symbol": "FXS",
                    "decimals": 18,
                    "FXS": "0x3432B6A60D23Ca0dFCa7761B7ab56459D9C964D0",
                    "VEFXS": "0xc8418aF6358FFddA74e09Ca9CC3Fe03Ca6aDC5b0",
                    "UNI_LP_FRAX_FXS": "0xE1573B9D29e2183B1AF0e743Dc2754979A40D237",
                    "FARMING_UNI_LP_FRAX_FXS": "0xda2c338350a0E59Ce71CDCED9679A3A590Dd9BEC"
                }
            }
        ],
        "addresses": [
            "0x36A87d1E3200225f881488E4AEedF25303FebcAe",
            "0x234D953a9404Bf9DbC3b526271d440cD2870bCd2",
            "0x6B175474E89094C44Da98b954EedeAC495271d0F",
            "0x22511ffA88433aB56495947b559dfB62B3d47633"
        ]
    }
}`
Seems to work!

**Comment by FortisFortuna** [comment]
Right, but if now A will also have 1000 votes (double voting). The end result should be 
A: 0
B: 1500
That is why in my original PR, I tried accounting for that.

**Comment by FortisFortuna** [comment]
> > Assume two voters, A and B
> > A has 1000, B has 500.
> > A then delegates to B.
> > Right now, B is showing 1000 instead of the sum, which should be 500 + 1000 = 1500, unless I am doing something wrong.
> 
> Ya you are correct It will return only delegated voting power, to add up original voting power too, you may have to add another strategy to get non delegated voting power, for example: https://hub.snapshot.org/api/spaces/balancer.eth they check both delegated and not delegated balance to vote
> 
> So in case both A and B vote, then the delegated voting power is not calculated for B. Only the vote of A will be calculated. The vote of B will be counted if A does not vote.

Isn't this exploitable? Someone could delegate one vote to a whale and kill his voting power

**Comment by FortisFortuna** [comment]
I created two test proposals:

1) **`frax-finance` and `delegation` strategies both enabled (Proposal 0x9e14)**
https://snapshot.org/#/fraxfinancegovernance.eth/proposal/0x9e143f8bfcb741c2b53eca4b031a16a91f92f6031f8baa79637a1dcb86a4fd19
fraxutilitywallet.eth (0x36A87d, delegator): 9.65k FXS
frax1.eth: (0x234D95, delegatee): 405.85 FXS

2) **`delegation` only (Proposal 0xb81e)**
https://snapshot.org/#/fraxfinancegovernance.eth/proposal/0xb81e938172f021f3f833bc18245b95d74535cf251deadee8f047ac8342e77bfd
fraxutilitywallet.eth (0x36A87d, delegator): 0 FXS (wouldn't let me vote)
frax1.eth: (0x234D95, delegatee): 9.65k FXS

I am wondering why in (2), it is not 9.65k + 405.85 for frax1.eth. I think this was the original issue.

**Comment by FortisFortuna** [comment]
Closed in favor of
https://github.com/snapshot-labs/snapshot-strategies/pull/299

**Pull request closed without merging**